Parse errors for a bunch of insertion modes with a word "head" in them

inikulin opened the pull request:
Parse errors for
```
12.2.6.4.3 The "before head" insertion mode
12.2.6.4.4 The "in head" insertion mode
12.2.6.4.5 The "in head noscript" insertion mode
12.2.6.4.6 The "after head" insertion mode
```
Preview: https://htmlparseerrorwg.github.io/html-build/output/multipage/parsing.html
\cc @HTMLParseErrorWG/core

inikulin committed:
Add misplaced-doctype error in before head mode

inikulin committed:
Add end-tag-without-matching-open-element error in before head mode

inikulin committed:
Add in head insertion mode errors

inikulin committed:
Add in head noscript insertion mode errors

inikulin committed:
Add after head insertion mode errors

inikulin commented:
@HTMLParseErrorWG/core ping

zcorpan commented on source:
```diff
@@ -99938,6 +99938,16 @@ dictionary <dfn>StorageEventInit</dfn> : <span>EventInit</span> {
     <tr>
+      <td><dfn data-x="parse-error-end-tag-without-matching-open-element">end-tag-without-matching-open-element</dfn>
+      <td><p>This error occurs if the parser encounters an <span data-x="syntax-end-tag">end
+      tag</span> that doesn't match any open element. The parser ignores such end tags.
```
The parser typically ignores such end tags. Exceptions are `</br>` and `</p>`.

zcorpan commented on source:
```diff
@@ -105596,7 +105609,9 @@ document.body.appendChild(text);
    <dd>
-    <p><span>Parse error</span>. Ignore the token.</p>
+    <p>This is an <span
+    data-x="parse-error-end-tag-without-matching-open-element">end-tag-without-matching-open-element</span>
```
This doesn't make sense for `<head>`. Maybe split up `<head>` and Any other end tag so they can have different error ids?

inikulin replied:
Was fixed in later commit

zcorpan commented on source:
```diff
@@ -99927,6 +99927,13 @@ dictionary <dfn>StorageEventInit</dfn> : <span>EventInit</span> {
     <tr>
+     <td><dfn data-x="parse-error-disallowed-content-in-noscript">disallowed-content-in-noscript</dfn>
+     <td><p>This error occurs if the parser encounters content in a <code>noscript</code> element
+     that is not allowed by it's content model. In such a case, the parser implicitly closes the
```
I think this is too overreaching to be accurate. For noscript in body in particular, you can put I think anything at all in noscript which violates the content model but does not close the noscript element. Maybe this needs to be limited to noscript in head? But even there you can have content model violations that doesn't close noscript (e.g., `<head><noscript><meta charset=utf-8>`). So maybe say something like "unexpected element in noscript in head".

inikulin replied:
Yeah, I've changed it to:
```
This error occurs if the parser encounters any content other than 
ASCII whitespace,  comments, a basefont element, a bgsound element, 
a link element, a meta element, a noframes element, or a style element 
within a noscript element. In this case, the parser implicitly closes
the noscript element and processes the content as if it was placed
after the noscript element in the markup.
```
in later commit. However, need to mention both in error code and text that it is related only to `noscript` in `head`.

zcorpan commented on source:
```diff
@@ -105687,15 +105703,21 @@ document.body.appendChild(text);
    <dd>
 
-    <!-- can't get here with an EOF and a fragment case -->
+    <!-- can't get here with a fragment case -->
```
Can you explain why this comment was wrong?

inikulin replied:
Because we definitely can, consider: `<!DOCTYPE html><head><noscript>`

inikulin replied:
I guess there was separate handling of EOF before, then it was refactored away, but comment left untouched.

zcorpan commented on source:
```diff
@@ -99859,6 +99859,36 @@ dictionary <dfn>StorageEventInit</dfn> : <span>EventInit</span> {
+     <span data-x="syntax-end-tag">end tag</span> and the <code>body</code> element's start tag or
+     any content that opens the <code>body</code> element if it's start tag is <span
+     data-x="syntax-tag-omission">omitted</span>. The parser treats such elements as child elements
+     of <code>head</code> element.</p>
```
Missing "the"

zcorpan commented on source:
```diff
@@ -99859,6 +99859,36 @@ dictionary <dfn>StorageEventInit</dfn> : <span>EventInit</span> {
+
+       <p>This will be parsed into:</p>
+
+       <ul class="domTree"><li class="t10">DOCTYPE: <code>html</code></li><li class="t1"><code>html</code><ul><li class="t1"><code>head</code><ul><li class="t1"><code>title</code><ul><li class="t3"><code>#text</code>: <span data-x="">The Answer</span></li></ul></li><li class="t1"><code>script</code> <span class="t2"><code class="attribute name" data-x="attr-id">src</code>="<code class="attribute  ...
```
Missing text nodes (with whitespace data) before and after the comment node, given the input markup.

inikulin replied:
I guess I'll better remove comment node from example: it will add too much visual noise to the resulting tree without having any crucial explanatory role.

zcorpan commented on source:
```diff
@@ -99859,6 +99859,36 @@ dictionary <dfn>StorageEventInit</dfn> : <span>EventInit</span> {
+      </div>
+
+      <p class="note">This error can be a sign of prematurely closed <code>head</code> element due
+      to a typo.</p>
```
What typo do you envision causing this? `</head>` instead of `<head>`? It seems to me this error is likely to happen if one does copy-and-paste and gets an extra `</head>` in the middle, or so. Or omitting the `<body>` tag and using a `<script>` as the first child of body, is a case I've run in to with the HTML spec itself actually. 😊
https://github.com/whatwg/wattsi/pull/48

inikulin replied:
> It seems to me this error is likely to happen if one does copy-and-paste and gets an extra </head> in the middle, or so

Yeah, something like this. But I guess I'll better remove this note, from the second sight it looks excessive.

zcorpan commented on source:
```diff
@@ -99928,16 +99958,29 @@ dictionary <dfn>StorageEventInit</dfn> : <span>EventInit</span> {
+     whitespace</span>, <span data-x="syntax-comments">comments</span>, a <code>basefont</code>
+     element, a <code>bgsound</code> element, a <code>link</code> element, a <code>meta</code>
+     element, a <code>noframes</code> element, or a <code>style</code> element within a
+     <code>noscript</code> element. In this case, the parser implicitly closes the
```
Add "in head"

zcorpan commented on source:
```diff
@@ -99928,16 +99958,29 @@ dictionary <dfn>StorageEventInit</dfn> : <span>EventInit</span> {
+    <tr>
+     <td><dfn data-x="parse-error-misplaced-start-tag-for-head-element">misplaced-start-tag-for-head-element</dfn>
+     <td><p>This error occurs if the parser encounters a <span data-x="syntax-start-tag">start
+     tag</span> whose tag name is "head" that is not the first start tag within the
```
Exception: `<html><html><head>` does not trigger this error. Maybe again say that `<head>` is encountered where it is not expected?

inikulin replied:
> where it is not expected

I find it a bit vague, since descriptions IMO should suggest how error can be fixed and such wording doesn't really clarify which position for `head`is allowed.

zcorpan commented:
I reviewed this commit-by-commit, so GitHub collapses away some of the comments.

inikulin commented on source:
```diff
@@ -105652,15 +105751,21 @@ document.body.appendChild(text);
    <dt>Any other end tag</dt>
    <dd>
-    <p><span>Parse error</span>. Ignore the token.</p>
+    <p>This is an <span
```
We shouldn't report this error for `A start tag whose tag name is one of: "head", "noscript"` case. Instead we need to report `misplaced-start-tag-for-head-element` error in case of "head" and new `nested-noscript-in-head` error for "noscript".

inikulin committed:
Add exceptions for end-tag-without-matching-open-element error. Limit scope of noscript error to <head>

inikulin committed:
Add missing article

inikulin committed:
Nits in abandoned-head-element-child error description

inikulin committed:
Clarify misplaced-start-tag-for-head-element

inikulin committed:
Process head in noscript in head using rules for head

inikulin committed:
nested-noscript-in-head error

inikulin commented:
@zcorpan I've addressed all issues you've found and one more that I've found recently (https://github.com/HTMLParseErrorWG/html/pull/22#discussion_r150069332). Can you take one more look, please?
